src/Umbraco.Cms.Integrations.SEO.SemrushTools/Umbraco.Cms.Integrations.SEO.SemrushTools/Controllers/SemrushController.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            return SemrushService.TryGetParameters(TokenDbKey, out TokenDto tokenDto) ? tokenDto : new TokenDto();
+        }
+
+        [HttpPost]
+        public void RevokeToken()
+        {
+            SemrushService.RemoveParameters(TokenDbKey);
+        }
+
+        [HttpPost]
+        public async Task<string> GetAccessToken([FromBody] AuthorizationRequestDto request)
+        {
+            var requestData = new Dictionary<string, string>
+            {
+                { "code", request.Code }
+            };
+
+            _client.DefaultRequestHeaders.Add("service_type", "SemrushAuthorization");
+
+            var response = await _client.PostAsync($"{AuthProxyBaseAddress}{AuthProxyTokenEndpoint}",
+                new FormUrlEncodedContent(requestData));
+            if (response.IsSuccessStatusCode)
+            {
+                var result = await response.Content.ReadAsStringAsync();
+
+                SemrushService.SaveParameters(TokenDbKey, result);
+
+                return result;
+            }
+
+            return "error";
+        }
+
+        [HttpPost]
+        public async Task<AuthorizationResponseDto> ValidateToken()
+        {
+            SemrushService.TryGetParameters(TokenDbKey, out TokenDto token);
+
+            if (string.IsNullOrEmpty(token.AccessToken)) return new AuthorizationResponseDto { IsExpired = true };
+
+            var response = await _client
+                .GetAsync(string.Format(SemrushKeywordsEndpoint, "phrase_related", token.AccessToken, "ping", "us"));
+
+            return new AuthorizationResponseDto
+            {
+                IsValid = response.StatusCode != HttpStatusCode.Unauthorized
+            };
+        }
+
+        [HttpPost]
+        public async Task<string> RefreshAccessToken()
+        {
+            SemrushService.TryGetParameters(TokenDbKey, out TokenDto token);
+
+            var requestData = new Dictionary<string, string> { { "refresh_token", token.RefreshToken } };
+
+            _client.DefaultRequestHeaders.Add("service_type", "SemrushReauthorization");
+
+            var response = await _client.PostAsync($"{AuthProxyBaseAddress}{AuthProxyTokenEndpoint}", new FormUrlEncodedContent(requestData));
+            if (response.IsSuccessStatusCode)
+            {
+                var result = await response.Content.ReadAsStringAsync();
+
+                SemrushService.SaveParameters(TokenDbKey, result);
+
+                return result;
+            }
+
+            return "error";
+        }
+
+        [HttpGet]
+        public async Task<RelatedPhrasesDto> GetRelatedPhrases(string phrase, int pageNumber, string dataSource, string method)
+        {
+            string cacheKey = $"{dataSource}-{method}-{phrase}";
+
+            if (CachingService.TryGetCachedItem(out var relatedPhrasesDto, cacheKey) && relatedPhrasesDto.Data != null)
+            {
+                relatedPhrasesDto.TotalPages = relatedPhrasesDto.Data.Rows.Count / 10;
```

Comment:
Should really move this `10` into a constant, so we could update it in one place and avoid a "magic number".

Reply:
Item addressed.